Ability to handle property Map<String, Object> in EntityManagerProvider.instance()

duschata opened the pull request:
Hi Rafael,
Unfortunately it is not possible to pass instances to the EntityManagerProvider.instance function. According to the specification, Persistence.createEntityManagerFactory() expects a map without describing exactly what type it is. For example, if you want to transfer your own validation instance (e.g. an instance which is managed by the BeanManager) to the factory  this is unfortunately not possible with the current EntityManagerProvider. So I modified the code a bit and as far as I can see and test it, there are no side effects with the PropertyResolutionUtil, which merges the system properties. Can you accept the change, please?
P.S.:You can see usages with instances passed to the factory here:
https://docs.jboss.org/hibernate/orm/5.3/userguide/html_single/Hibernate_User_Guide.html#_bean_validation_options

Kind regards and greetings from Darmstadt/Germany

Tom

rmpestano commented:
Hey @duschata, thank you for your contribution!


Can you do the same with the one in core? I really don't remember why we have 2 EmPtoviders, if you can investigate and try to have only one in core it would be amazing!

Cheers!

rmpestano commented:
> I really don't remember why we have 2 EmProviders, if you can investigate and try to have only one in core it would be amazing!

I just had a quick look and seems it's because the one in core imolements class rule:
https://github.com/database-rider/database-rider/blob/e72c4d5bf41dd605727819bc6ab41051bf2360b2/rider-core/src/main/java/com/github/database/rider/core/util/EntityManagerProvider.java#L22

What we could do is to have a EntityManagerProviderRule which uses EntityManagerProvider and then we would be able to have only one EntotyManagerProvider I think

duschata commented:
Hi Rafael! Thanks for your response. Please, give me a little bit of time. I think I've found another issue in the context of  the EntityManagerProvider and I will investigate (and hopefully fix) this first.
Cheers

rmpestano commented:
Looks like we're good to merge, right? Thank you @duschata!

duschata commented:
I'm not sure:

1. All tests are running localy but there where failures at the github CICD. 
2. We have a "breaking change" because the EntityManagerProvider in the junit5 namespace does not exist anymore. Those who use this one will get compile errors. We can not work wiht extends because there are factory constructors in the class.
I would suggest to copy the EntityManagerProvider from the core package 1:1 into the junit5 package and mark this one as deprecated and remove it in a later release.

rmpestano commented:
Hi again.

> All tests are running localy but there where failures at the github CICD.

Yes that is happening with the leakHunter test and I'm not sure why

> I would suggest to copy the EntityManagerProvider from the core package 1:1 into the junit5 package and mark this one as deprecated and remove it in a later release.

That's a good point, let's do that or maybe call core impl on junit5 emProvider methods?
I have plans to release Rider 2.0 until the end of the year (Java 8 as baseline and use value comparers in expected dataset) and then we could remove the emProvider from junit5

duschata commented:
Hi Rafael,
Ups, the test are working localy except one: 
https://github.com/duschata/database-rider/blob/emfactory_props_as_objects/rider-junit5/src/test/java/com/github/database/rider/junit5/DBRiderSpringDataSourceIt.java#L65
can you please take a closer look what happens there?
The backwards compatibility was a litte bit tricky, because DBUnitExtension and JUnit5RiderTestContext have reference to the EntityManagerProvider in the junit5 package. So I copied them to a new 'incubating' package to test the new behavior and preserve also the old. I hope this does not confuse the users. Sorry, I saw no other choice to solve this.
Kind regards and happy weekend,
Tom

rmpestano commented:
> Hi Rafael, Ups, the test are working localy except one: https://github.com/duschata/database-rider/blob/emfactory_props_as_objects/rider-junit5/src/test/java/com/github/database/rider/junit5/DBRiderSpringDataSourceIt.java#L65 can you please take a closer look what happens there? The backwards compatibility was a litte bit tricky, because DBUnitExtension and JUnit5RiderTestContext have reference to the EntityManagerProvider in the junit5 package. So I copied them to a new 'incubating' package to test the new behavior and preserve also the old. I hope this does not confuse the users. Sorry, I saw no other choice to solve this. Kind regards and happy weekend, Tom


Hi @duschata, actually what I was thinking of, was to call EmProvider from core inside the one on junit5 module, have you tried that? if that doesn't work I'd keep as before, 2 emProviders with your initial implementation on this PR, WDYT?

duschata commented:
Hi Rafael
As far as I understand you would like some kind of deligator pattern for the EntityManagerProvider in the junit5 package. There are currently 19 public functions there, which would delegate their calls to the EntityManagerProvider in the core package, which is quite a lot. When calling instance (...) from the junit5 package, an EntityManagerProvider instance from the core package would be returned.
That would lead to the following construction, if we call EntityManagerProvider from the junit5 package and need a var for later usage:
```
com.github.database.rider.core.util.EntityManagerProvider test = EntityManagerProvider.newInstance ("test"); 
```
Some user code wouldn't compile either, would it?
If I get you wrong, please contact me again.

BTW: at the moment I'm still working on the
```
private static Map <String, Object> overridingProperties;
```
Unfortunately, this variable is static which means that a call from
```
instance (pu: String) 
```
after a call from 
```
instance (pu: String, props: Map)
```
takes over the properties
So please don't merge the PR yet. I'll contact you later

Greetings
Tom

rmpestano commented:
> As far as I understand you would like some kind of deligator pattern for the EntityManagerProvider in the junit5 package. There are currently 19 public functions there, which would delegate their calls to the EntityManagerProvider in the core package, which is quite a lot. When calling instance (...) from the junit5 package, an EntityManagerProvider instance from the core package would be returned.

Hi again, yes true and that would break current users expecting junit5 provider, that would be possible if we had a common interface but don't ;(.

I would say, don't spend too much time on something we might remove in rider 2.0 so I'd keep as we have now in master and would add your fix in both providers.

Thank you a lot for putting effort into this!

duschata commented:
Ok, thanks. But please wait a moment, I'm still working on the overridingProperties fix...

greetings Tom

duschata commented:
done... I've fixed the overridingProperties and added a removeInstance method in order to absolutly ensure that a new instance of  the emf could be created, if sombody like to do this. Eventually this class should be refactored in the 2.0 release.
greetings...

Tom
